src/log-store/src/raft_engine/log_store.rs

```diff
@@ -116,22 +116,71 @@ impl RaftEngineLogStore {
         )
     }
 
-    /// Checks if entry does not override the min index of namespace.
-    fn check_entry(&self, e: &EntryImpl) -> Result<()> {
-        if cfg!(debug_assertions) {
+    /// Converts entries to `LogBatch` and checks if entry ids are valid.
+    /// Returns the `LogBatch` converted along with the last entry id
+    /// to append in each namespace(region).
+    fn entries_to_batch(
+        &self,
+        entries: Vec<EntryImpl>,
+    ) -> Result<(LogBatch, HashMap<NamespaceId, EntryId>)> {
+        // Records the last entry id for each region's entries.
+        let mut entry_ids: HashMap<NamespaceId, BTreeSet<EntryId>> =
```

Comment:
We don't need to store entry ids in the `BTreeSet`. We can store the last entry id in the map
- if the entry is vacant, check `first_in_batch`
- if the entry is occupied, check `current entry id == entry id in map + 1` and update the entry id in map.

Reply:
> We don't need to store entry ids in the `BTreeSet`. We can store the last entry id in the map
> 
> * if the entry is vacant, check `first_in_batch`
> * if the entry is occupied, check `current entry id == entry id in map + 1` and update the entry id in map.

Maybe change the `entry_ids` to `HashMap<NamespaceId, (EntryId, EntryId)>`. 
For the value part, the left is the first entry id of region in batch while the right side is the current max entry id. So that we can check if :
- First entry id in batch equals first_index+1
- Current entry id equals current max entry id +1